### ISSUE TITLE
New Rule: Ordering of JOIN types

### ISSUE THREAD

**Issue by jpy-git** [sqlfluff/sqlfluff#2391]
From Kickstarter style guide: https://gist.github.com/fredbenenson/7bb92718e19138c20591#join

`Begin with INNER JOINs and then list LEFT JOINs, order them semantically, and do not intermingle LEFT JOINs with INNER JOINs unless necessary:`

GOOD:

```sql
INNER JOIN ksr.backings AS backings ON ...
INNER JOIN ksr.users AS users ON ...
INNER JOIN ksr.locations AS locations ON ...
LEFT JOIN ksr.backer_rewards AS backer_rewards ON ...
LEFT JOIN ...
```

BAD:

```sql
LEFT JOIN ksr.backer_rewards AS backer_rewards ON backings
INNER JOIN ksr.users AS users ON ...
LEFT JOIN ...
INNER JOIN ksr.locations AS locations ON ...
```

Quite like the idea of grouping the join types together.

We'd need to choose an ordering to generalise this further so as a first pass let's go with:
1. INNER JOIN/JOIN
2. LEFT JOIN
3. RIGHT JOIN
4. FULL JOIN
5. CROSS JOIN

**Comment by barrywhart** [comment]
Random observation: Assuming this rule generates fixes, it will reorder code in a manner similar to L034. For a while, we were encountering a **lot** of `sqlfluff fix` issues with L034 and templating. I _think_ we're in excellent shape now and won't see these kinds of issues anymore. 🤞

**Comment by jpy-git** [comment]
yeah it'll be a good test! I already have a fairly good idea of how this rule should work so will probably work on this one next 👍

**Comment by jpy-git** [comment]
Do you agree with my proposed ordering?

**Comment by barrywhart** [comment]
The ordering makes sense to me.

I'm unsure what it means to:
>order them semantically

Is that something SQLFluff can detect and/or enforce?

Related: I think someone once proposed a rule to discourage the use of `RIGHT JOIN`, but I can't find it. Was just thinking that's a somewhat related idea to this one.

**Comment by jpy-git** [comment]
So I think ordering semantically means order them in terms of importance/relevance to the query logic, which is obviously something SQLFluff can't infer. That's why I'm suggesting we just generalise to ordering by JOIN type (within the same JOIN type they can just be ordered by how the originally appeared) 😄 

Yeah I think the forbid `RIGHT JOIN` was one I suggested a while back (will have a look for where that was). That would be a separate rule though. Some people might want to allow RIGHT JOINs and as such this rule should be able to include them in the ordering. Will make a separate issue for forbidding RIGHT JOINS.
